ENH: Add return_back_azimuth kwarg to fwd, inv, fwd_intermediate, inv_intermediate

idanmiara opened the pull request:
https://github.com/pyproj4/pyproj/issues/1163

 - [X] Closes #1163
 - Tests added
   - [ ] inv (also test warning)
   - [ ] inv_intermediate (also test warning)
   - [ ] fwd (also test warning)
   - [ ] fwd_intermediate (also test warning)
   - [ ] reverse_azimuth
   - [ ] reverse_azimuth_arr
 - Fully documented
   - [x] `history.rst`
   -  `api/*.rst` for new API
     - [ ] reverse_azimuth
     - [ ] reverse_azimuth_arr

snowman2 commented on pyproj/utils.py:
```diff
@@ -151,3 +152,40 @@ def _convertback(data_type: DataType, inx: Any) -> Any:
     return inx
+
+
+def reverse_azimuth(azi: float, radians: bool = False) -> float:
```
I would say these functions are a better fit for `geod.py` than a generic pyproj utility.

snowman2 replied:
It looks like they are only used for testing. Thoughts about moving them to the testing module instead?

snowman2 replied:
It may be more performant to move them to `_geod.pyx` and import into `_geod.py`.

idanmiara replied:
> It may be more performant to move them to `_geod.pyx` and import into `geod.py`.

makes sense, would you mind doing that after this merge as I don't have cython installed and I don't have more time to invest in this...

idanmiara commented:
@snowman2 I think this is ready. 
I could reorganize the commits if you approve this for merging.

snowman2 reviewed (approved):
Thanks @idanmiara :+1:

snowman2 commented:
I am not going to merge it quite yet. 3.4.1 has not been released yet and this is targeted for 3.5. 3.4.1 may be replaced by 3.5, but not sure yet.

snowman2 commented on pyproj/geod.py:
```diff
@@ -240,9 +240,17 @@ def __init__(self, initstring: Optional[str] = None, **kwargs) -> None:
+        az: Any,
+        dist: Any,
+        radians: bool = False,
+        return_back_azimuth: bool = True,
```
I have been thinking about this and am thinking it would be a good idea to warn users of
the upcoming change.

This change can initially be released with:
```
return_back_azimuth: Optional[bool] = None,
```
and 
```python
if return_back_azimuth is None:
    return_back_azimuth = False
    warnings.warn(
        "The forward azimuth is currently being returned by default. "
        "In the future, the back azimuth will be returned. To opt-in "
        "to the new behavior, set return_back_azimuth=True. "
        "Otherwise, to avoid breaking changes, set return_back_azimuth=False "
        "for pyproj 3.5+."
    )
```
Then, in a future release, the default can be changed to True.

Thoughts?

snowman2 replied:
Another idea is to change the default and raise a warning similar to the above if the user has not selected True or False so they know that they are getting the back azimuth now.

idanmiara replied:
> Another idea is to change the default and raise a warning similar to the above if the user has not selected True or False so they know that they are getting the back azimuth now.

Yes, sounds good.

snowman2 replied:
Great. @idanmiara, is this something you have time to add to this PR?

idanmiara replied:
yes. added and rebased.

snowman2 commented:
Thanks @idanmiara for the changes. We're getting closer. I updated the initial comment on the issue with a checklist of testing & documentation updates that would be good to add. Let me know if you have time to add those.

idanmiara commented:
> Thanks @idanmiara for the changes. We're getting closer. I updated the initial comment on the issue with a checklist of testing & documentation updates that would be good to add. Let me know if you have time to add those.

All good now?

snowman2 commented on pyproj/geod.py:
```diff
@@ -1067,3 +1115,49 @@ def __eq__(self, other: Any) -> bool:
+try:
+    import numpy as np  # type: ignore  # pylint: disable=import-outside-toplevel
+
+    def reverse_azimuth(azi: Any, radians: bool = False) -> Any:
```
This function does not appear to behave the same as the one below and changes in place based on the existence of numpy. You should be able to re-write the function to work on a numpy array without needing numpy. I recommend using that approach as it is easier to test and ensure that the behavior is consistent with or without numpy.

snowman2 replied:
This may be useful:
- https://docs.python.org/3/library/math.html#math.copysign
- https://stackoverflow.com/a/1986776/11622836

idanmiara replied:
Yes, I thought about using `copysign`, but how would you apply this to a numpy array?
I couldn't find a way to write that function that will work on numpy array and on scalars without importing numpy.

snowman2 replied:
Some functions in the math module work on numpy arrays, but that one unfortunately doesn't. You will likely have to do a workaround for numpy arrays. Something like this:

```python
def sign(data):
    data_sign = data.copy()
    neg = data < 0
    data_sign[neg] = -1
    data_sign[~neg] = 1
    return data_sign
```

idanmiara replied:
This function would work on numpy array but not on scalars, so if you write two versions of the functions, why not use the built `sign` method as I did?

snowman2 replied:
It is an equivalent amount of code and easier to test/maintain if you don't use numpy.

snowman2 replied:
Actually, I have an idea that I think we both will like. Will share later when I am not on my phone.

idanmiara replied:
I see that numpy has also `copysign` so maybe now it's better.

snowman2 replied:
My main goal is to remove numpy from the core codebase. Otherwise, the tests and CI will need to be refactored to support both testing with and without numpy. The added complexity doesn't sound fun to maintain.

Here is an alternative approach: https://github.com/talos-gis/pyproj/pull/1

It uses cython instead of numpy and it supports all of the input scalar/array types that the geodesic routines support. This is easier to test and handles more input types in a performant way when numpy isn't present.

idanmiara replied:
Great, thanks for looking into this 👍

snowman2 commented:
This looks great - thanks @idanmiara :+1:. I think this is ready to go. Would you like to clean up the commit history or would you mind if I squashed on merge?

idanmiara commented:
> This looks great - thanks @idanmiara 👍. I think this is ready to go. Would you like to clean up the commit history or would you mind if I squashed on merge?

Thanks for your help!
I've cleaned the commit history.
